### ISSUE TITLE
CloudTable.execute(query) on WADMetrics table takes forever to return

### ISSUE THREAD

**Issue by gadagip** [Azure/azure-storage-java#103]
I'm trying to use the azure-storage-sdk to obtain VM metrics using the storage account associated with Diagnostics. I get the Metric table name using the GET /metricDefinitions API. Using this table name, I'm getting the metric values. Here's my code:

    // Retrieve storage account from connection-string.
	CloudStorageAccount storageAccount = CloudStorageAccount.parse(storageConnectionString);
	// Create the table client.
	CloudTableClient tableClient = storageAccount.createCloudTableClient();
	
	TableRequestOptions defaultRequestOptions = new TableRequestOptions(tableClient.getDefaultRequestOptions());=
	 defaultRequestOptions.setTimeoutIntervalInMs(10);
	tableClient.setDefaultRequestOptions(defaultRequestOptions);
				
	CloudTable cloudTable = tableClient.getTableReference(TABLE);
	String partitionVal = <PARTITION_KEY>;
	String partitionFilter = TableQuery.generateFilterCondition(PARTITION_KEY, QueryComparisons.EQUAL, partitionVal);
	
	long endTimeMicros = getNowMicrosUtc();
	Date timeStamp = new Date(TimeUnit.MICROSECONDS.toMillis(endTimeMicros) - TimeUnit.MINUTES.toMillis(1));
	
	String timestampFilter = TableQuery.generateFilterCondition(TIMESTAMP, 
	QueryComparisons.GREATER_THAN_OR_EQUAL, timeStamp);
	
	String partitionAndTimestampFilter = TableQuery.combineFilters(partitionFilter, Operators.AND, timestampFilter);
	
	for (String metric : metrics) {
		String rowFilter = TableQuery.generateFilterCondition(ROW_KEY, QueryComparisons.EQUAL, metric);
		String combinedFilter = TableQuery.combineFilters(partitionAndTimestampFilter, Operators.AND, rowFilter);
		TableQuery<DynamicTableEntity> partitionQuery = TableQuery.from(DynamicTableEntity.class).where(combinedFilter);
					ResultSegment<DynamicTableEntity> entities = cloudTable.executeSegmented(partitionQuery, null);
					for (DynamicTableEntity entity : entities.getResults()) {
						HashMap<String, EntityProperty> properties = entity.getProperties();
						System.out.println("RowKey: " + entity.getRowKey());
						System.out.println(entity.getTimestamp());
						for (String key : properties.keySet()) {
							System.out.println(key + " --> " + properties.get(key).getValueAsString());
						}
						System.out.println("==========================================================");
						break;
					}

The issue I'm facing is the table.executeSegmented() or table.execute() takes a long time to return. Sometimes, it's quick. But a lot of the times, takes more than a minute to return with data.

Could you please look into it?

**Comment by pemari-msft** [comment]
It's impossible to tell what's dominating your latency from just the client library calls. Can you please try enabling [Client Side Logs](https://msdn.microsoft.com/en-us/library/azure/dn782844.aspx) and [Server Analytics Logs](https://msdn.microsoft.com/en-us/library/azure/dn782840.aspx) which will tell you where the latency is (client side or server side), or if there's throttling that is causing retries?

If you're seeing server latency and not End To End Latency in your storage analytics logs, please do reach out to us at `ascl [at] microsoft [dot] com` with your account name and specific slow request IDs. We'll be able to see what is happening and give you a detailed explanation. Thanks!

**Comment by gadagip** [comment]
@pemari-msft After enabling both client side and server side logs, I don't see any of those logs in my console window.

This is what I did to enable client side and server side logs:
```
// Enable Server side logging
CloudQueueClient queueClient = storageAccount.createCloudQueueClient();
ServiceProperties serviceProperties = queueClient.downloadServiceProperties();

LoggingProperties logginProperties = new LoggingProperties();
logginProperties.setLogOperationTypes(EnumSet.of(LoggingOperations.WRITE, LoggingOperations.DELETE, LoggingOperations.READ));
logginProperties.setRetentionIntervalInDays(2);
serviceProperties.setLogging(logginProperties);
queueClient.uploadServiceProperties(serviceProperties);

// Enable Client side logging
OperationContext.setLoggingEnabledByDefault(false);
OperationContext ctx = new OperationContext();
ctx.setLoggingEnabled(true);
```

**Comment by asorrin-msft** [comment]
@gadagip For client-side logs, did you add the slf4j JAR files to the build path of your project?  And, just to check, you're using the ctx as the input to the cloudTable.executeSegmented call?
For the server-side logs, these are stored on the service, you'll need to download them separately:
https://msdn.microsoft.com/en-us/library/azure/dn782840.aspx#DownloadingStorageLogginglogdata

**Comment by gadagip** [comment]
@asorrin-msft @pemari-msft 

I added the slf4j jars and from the logs I can see that there is a 5 second delay between sending the request and receiving the response.
```
2016-08-01 10:58:33 INFO  root:121 - {03db951a-c48a-4c40-8386-dcc077cfd25c}: {Starting operation.}
2016-08-01 10:58:33 INFO  root:153 - {03db951a-c48a-4c40-8386-dcc077cfd25c}: {Starting operation with location 'PRIMARY' per location mode 'PRIMARY_ONLY'.}
2016-08-01 10:58:33 INFO  root:153 - {03db951a-c48a-4c40-8386-dcc077cfd25c}: {Starting request to 'http://enumtestdonotdelete5119.table.core.windows.net/WADMetricsPT1MP10DV2S20160724?$filter=%28%28PartitionKey%20eq%20%27%3A002Fsubscriptions%3A002F1fe92c6a%3A002D2dda%3A002D4fd8%3A002D8941%3A002D6b6ba9da3286%3A002FresourceGroups%3A002FEnumTest%3A002DDoNotDelete%3A002Fproviders%3A002FMicrosoft%3A002ECompute%3A002FvirtualMachines%3A002FEnumTestVM%3A002DDoNotDelete%27%29%20and%20%28Timestamp%20ge%20datetime%272016-08-01T17%3A57%3A02.581Z%27%29%29%20and%20%28CounterName%20eq%20%27%5CMemory%5CUsedMemory%27%29&timeout=0' at 'Mon, 01 Aug 2016 17:58:33 GMT'.}
2016-08-01 10:58:33 INFO  root:121 - {03db951a-c48a-4c40-8386-dcc077cfd25c}: {Waiting for response.}
2016-08-01 10:58:38 INFO  root:131 - {03db951a-c48a-4c40-8386-dcc077cfd25c}: {Response received. Status code = '200', Request ID = '23af5a7c-0002-0035-4e1e-ec83bc000000', Content-MD5 = 'null', ETag = 'null', Date = 'Mon, 01 Aug 2016 17:58:29 GMT'.}
2016-08-01 10:58:38 INFO  root:121 - {03db951a-c48a-4c40-8386-dcc077cfd25c}: {Processing response headers.}
2016-08-01 10:58:38 INFO  root:121 - {03db951a-c48a-4c40-8386-dcc077cfd25c}: {Response headers were processed successfully.}
2016-08-01 10:58:38 INFO  root:121 - {03db951a-c48a-4c40-8386-dcc077cfd25c}: {Processing response body.}
2016-08-01 10:58:38 INFO  root:121 - {03db951a-c48a-4c40-8386-dcc077cfd25c}: {Response body was parsed successfully.}
2016-08-01 10:58:38 INFO  root:121 - {03db951a-c48a-4c40-8386-dcc077cfd25c}: {Operation completed.}
```

Seems like a server side issue?

**Comment by pemari-msft** [comment]
Looks like you're doing a partition scan that returns one row. It takes time to scan the entire partition for the row so that's the time you're seeing. This case looks expected to me.

**Comment by solvingJ** [comment]
@gadagip  It looks like your problem might be related to your row key filter. Your wireshark does not show that a rowkey parameter was included in the final query. Looking at the line below, it doesn't look like you're setting it to an appropriate value.  It seems like it's just being discarded. 

`String rowFilter = TableQuery.generateFilterCondition(ROW_KEY, QueryComparisons.EQUAL, metric);`

I'm still struggling with setting a rowkey filter myself, but the following is taken from the .NET library which queries VM metrics: 

`        private static TableQuery GenerateMetricTimestampQuery(string partitionKey, DateTime startTime, DateTime endTime)
        {
            return GenerateMetricQuery(
                partitionKey,
                (DateTime.MaxValue.Ticks - endTime.Ticks + 1).ToString("D19") + "__",
                (DateTime.MaxValue.Ticks - startTime.Ticks).ToString("D19") + "__");
        }

        // Creates a TableQuery object which filters entities to a particular partition key and the given row key range
        private static TableQuery GenerateMetricQuery(string partitionKey, string startRowKey, string endRowKey)
        {
            string partitionFilter = TableQuery.GenerateFilterCondition("PartitionKey", QueryComparisons.Equal, partitionKey);
            string rowStartFilter = TableQuery.GenerateFilterCondition("RowKey", QueryComparisons.GreaterThanOrEqual, startRowKey);
            string rowEndFilter = TableQuery.GenerateFilterCondition("RowKey", QueryComparisons.LessThan, endRowKey);
            
            return new TableQuery()
            {
                FilterString = TableQuery.CombineFilters(partitionFilter, TableOperators.And, TableQuery.CombineFilters(rowStartFilter, TableOperators.And, rowEndFilter))
            };
        }`


@pemari-msft - can you perhaps offer any guidance about how to use Java Dates and convert them to a suitable value for the "RowKey" Filter?  Java dates don't seem to have Ticks, so I don't know what the equivalent is.

**Comment by solvingJ** [comment]
@gadagip  

The creation of timestamps needed for specifying RowKey is much more complicated than your example.  Fortunatley, Andreas from Stackoverflow gives us a working solution for generating row key filters!

See below: 
[http://stackoverflow.com/questions/39112101/java-8-time-equivalent-of-net-datetime-maxvalue-ticks](url)

**Comment by gadagip** [comment]
@solvingJ Creating of "RowKey" was a big issue while retrieving values from Azure Tables. Which is why, I did not set any "RowKey". The variable ROW_KEY in my example is misleading because all it points to is the CounterName, which is the name of the metric.

Regardless, I noticed there are 2 "RowKey" that are generated. One with the ticks suffixed and the other with ticks prefixed. I can give Andreas' solution to generate the ticks a try and see if that helps.
